app/src/full/java/io/homeassistant/companion/android/sensors/HealthConnectSensorManager.kt

```diff
@@ -0,0 +1,168 @@
+                    timeRangeFilter = TimeRangeFilter.between(
+                        LocalDateTime.of(LocalDate.now(), LocalTime.MIDNIGHT),
+                        LocalDateTime.of(LocalDate.now(), LocalTime.now())
+                    )
+                )
+            )
+        }
+        totalCaloriesBurnedRequest[TotalCaloriesBurnedRecord.ENERGY_TOTAL]?.let {
+            onSensorUpdated(
+                context,
+                totalCaloriesBurned,
+                it.inKilocalories,
+                totalCaloriesBurned.statelessIcon,
+                attributes = mapOf("endTime" to LocalDateTime.of(LocalDate.now(), LocalTime.now()).toInstant(ZoneOffset.UTC))
+            )
+        }
+    }
+
+    private fun updateWeightSensor(context: Context, healthConnectClient: HealthConnectClient) {
+        val weightRequest = ReadRecordsRequest(
+            recordType = WeightRecord::class,
+            timeRangeFilter = TimeRangeFilter.between(
+                Instant.now().minus(30, ChronoUnit.DAYS),
+                Instant.now()
+
+            ),
+            ascendingOrder = false,
+            pageSize = 1
+        )
+        val response = runBlocking { healthConnectClient.readRecords(weightRequest) }
+        onSensorUpdated(
+            context,
+            weight,
+            response.records.last().weight.inGrams,
+            weight.statelessIcon,
+            attributes = mapOf("date" to response.records.last().time)
+        )
+    }
+
+    private fun updateActiveCaloriesBurnedSensor(context: Context, healthConnectClient: HealthConnectClient) {
+        val activeCaloriesBurnedRequest = ReadRecordsRequest(
+            recordType = ActiveCaloriesBurnedRecord::class,
+            timeRangeFilter = TimeRangeFilter.between(
+                Instant.now().minus(30, ChronoUnit.DAYS),
+                Instant.now()
+
+            ),
+            ascendingOrder = false,
+            pageSize = 1
+        )
+        val response = runBlocking { healthConnectClient.readRecords(activeCaloriesBurnedRequest) }
+        if (response.records.isEmpty()) {
+            return
+        }
+        onSensorUpdated(
+            context,
+            activeCaloriesBurned,
+            response.records.last().energy.inKilocalories,
+            activeCaloriesBurned.statelessIcon,
+            attributes = mapOf("endTime" to response.records.last().endTime)
+        )
+    }
+
+    override suspend fun getAvailableSensors(context: Context): List<SensorManager.BasicSensor> {
+        return if (hasSensor(context)) {
+            listOf(weight, activeCaloriesBurned, totalCaloriesBurned)
+        } else {
+            emptyList()
+        }
+    }
+
+    override fun hasSensor(context: Context): Boolean {
+        return SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE
+    }
+
+    override fun checkPermission(context: Context, sensorId: String): Boolean {
+        val healthConnectClient = HealthConnectClient.getOrCreate(context)
```

Comment:
this is also called during `requestSensorUpdates` maybe the variable should be shared instead of recreated twice during a sensor update?